### PR TITLE
[FEATURE] Added BlockedUsers Management

### PR DESCRIPTION
**Pull request by TheClashster**
Please prefix your pull request with one of the following: **[FEATURE]** **[FIX]** **[IMPROVEMENT]**.

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/sample-platform/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [x] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used the project.
- [x] I have used the project briefly.
- [ ] I have used the project extensively, but have not contributed previously.
- [ ] I am an active contributor to the project.

---

Benevolent dictator don't be mad plis :)

### CODE REVIEW

File: mod_ci/controllers.py [hunk 1]
```diff
@@ -903,6 +903,63 @@ def show_maintenance():
...
+@login_required
+@check_access_rights([Role.admin])
+@template_renderer()
+def blocked_users():
+    return {
+        'blocked_users': BlockedUsers.query.order_by(BlockedUsers.userID)
+    }
+
+
+@mod_ci.route('/blocked_users')
```
Comment (saurabhshri) [inline mod_ci/controllers.py]: Multiple definitions for a single route? You don't need to write function defs here.

`@mod_ci.route('/blocked_users')` for this you'll write exactly what should happen if someone opens this path. Check other snippets in the same file to get an idea.

File: mod_ci/controllers.py [hunk 2]
```diff
@@ -903,6 +905,63 @@ def show_maintenance():
...
+@template_renderer('ci/blocked_users.html')
+def username(u):
+    userID = str(u)
+    api_url = requests.get('https://api.github.com/user/' + userID)
+    userdata = api_url.json()
+    username = userdata['login']
+    return username
+
+
+app.jinja_env.globals.update(username=username)
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: The app scope is not defined. I'd prefer the filters to be grouped together.

File: mod_ci/controllers.py [hunk 3]
```diff
@@ -43,6 +45,8 @@
 
 mod_ci = Blueprint('ci', __name__)
 
+app = Flask(__name__)
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: This is a big no-no. There must be only one app defined, and that's in run.py

File: mod_ci/controllers.py [hunk 4]
```diff
@@ -903,6 +907,63 @@ def show_maintenance():
...
+@check_access_rights([Role.admin])
+@template_renderer()
+def blocked_users():
+    return {
+        'blocked_users': BlockedUsers.query.order_by(BlockedUsers.userID)
+    }
+
+
+@mod_ci.route('/blocked_users')
+@login_required
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: That def, if it's used in a template, doesn't need any of these decorators.

File: mod_ci/forms.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from flask_wtf import Form
+from wtforms import IntegerField, SubmitField, StringField
+from wtforms.validators import DataRequired
+
+
+class AddUsersToBlacklist(Form):
+    userID = IntegerField('User ID', [DataRequired(
```
Comment (canihavesomecoffee) [inline mod_ci/forms.py]: This can be put onto one line. We allow up to 120 chars :)

File: mod_ci/forms.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        message='GitHub User ID not filled in')])
+    comment = StringField('Comment')
+    submit = SubmitField('Add User')
+
+
+class RemoveUsersFromBlacklist(Form):
+    userID = IntegerField('User ID', [DataRequired(
```
Comment (canihavesomecoffee) [inline mod_ci/forms.py]: Same here

File: templates/ci/blocked_users.html [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                    <td>{{ u.comment }}</td>
+                </tr>
+            {% endfor %}
+            </tbody>
+        </table>
+    </div>
+    <div class="row">
+        <h1>Add Users</h1>
+    </div>
+    <form method="post" name="registrationForm" id="registrationForm" action="{{ url_for('.add_user_to_blacklist') }}">
```
Comment (canihavesomecoffee) [inline templates/ci/blocked_users.html]: Having duplicate name's and id's is not a good idea. Please correct.

File: mod_ci/controllers.py [hunk 8]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+    api_url = requests.get('https://api.github.com/user/' + user_id)
+    userdata = api_url.json()
+    username = userdata['login']
+    return username
+
+
+app.jinja_env.globals.update(username=username)
+
+
+def add_user_to_blacklist():
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: This will need decorators if it is referenced by an url_for.

File: mod_ci/controllers.py [hunk 9]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+        blocked_user = BlockedUsers(form.userID.data, form.comment.data)
+        if blocked_user == BlockedUsers.query.filter_by(userID=form.userID.data).first():
+            flash('User already blocked.')
+            return redirect(url_for('ci.blocked_users'))
+        else:
+            g.db.add(blocked_user)
+            g.db.commit()
+
+
+def remove_user_from_blacklist():
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Same.

File: mod_ci/controllers.py (hunk 1 above)
Comment (TheClashster) [reply mod_ci/controllers.py]: Done, correctly I think 🤦‍♂️

File: templates/ci/blocked_users.html [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            <div class="medium-6 columns">
+                {{ macros.render_field(form.comment) }}
+            </div>
+            <div class="medium-12 columns">
+                {{ macros.render_field(form.submit) }}
+            </div>
+    <div class="row">
+        <h1>Remove Blocked Users</h1>
+    </div>
+    <form method="post" name="registrationForm" id="registrationForm" action="{{ url_for('.remove_user_from_blacklist') }}">
```
Comment (canihavesomecoffee) [inline templates/ci/blocked_users.html]: Same here

File: mod_ci/controllers.py (hunk 2 above)
Comment (canihavesomecoffee) [reply mod_ci/controllers.py]: That is in the run.py.

File: mod_ci/controllers.py [hunk 11]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+app.jinja_env.globals.update(username=username)
+
+
+def add_user_to_blacklist():
+    form = AddUsersToBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers(form.userID.data, form.comment.data)
+        if blocked_user == BlockedUsers.query.filter_by(userID=form.userID.data).first():
+            flash('User already blocked.')
+            return redirect(url_for('ci.blocked_users'))
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: you can use a `.blocked_users`, as you're in the module.

File: mod_ci/controllers.py [hunk 12]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+            g.db.commit()
+
+
+def remove_user_from_blacklist():
+    form = RemoveUsersFromBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers.query.filter_by(userID=form.userID.data).first()
+        if blocked_user is None:
+            flash('No such user in Blacklist')
+            return redirect(url_for('ci.blocked_users'))
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Same remark as with other redirect.

File: mod_ci/controllers.py [hunk 13]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+        else:
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: If you have a return in the if, you don't need this else; just do no else and remove the indendation.

File: mod_ci/controllers.py [hunk 14]
```diff
@@ -903,6 +907,51 @@ def show_maintenance():
...
+
+
+def add_user_to_blacklist():
+    form = AddUsersToBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers(form.userID.data, form.comment.data)
+        if blocked_user == BlockedUsers.query.filter_by(userID=form.userID.data).first():
+            flash('User already blocked.')
+            return redirect(url_for('ci.blocked_users'))
+        else:
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Else branch is unnecessary if you have a return in the if.

File: mod_ci/controllers.py [hunk 15]
```diff
@@ -538,15 +539,15 @@ def start_ci():
...
-                user_id = payload['pull_request']['user']['id']
-                if BlockedUsers.query.filter(BlockedUsers.userID == user_id).first() is not None:
-                    g.log.critical("User Blacklisted")
-                    gh_commit.post(
-                        state=Status.ERROR,
-                        description="CI start aborted. You may be blocked from accessing this functionality",
-                        target_url=url_for('home.index', _external=True)
-                    )
-                    return 'ERROR'
+                if BlockedUsers.query.filter(BlockedUsers.userID == payload['pull_request'][
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Please undo this change, or rebase against master to have it.

File: mod_ci/controllers.py [hunk 16]
```diff
@@ -866,6 +867,44 @@ def show_maintenance():
...
+@check_access_rights([Role.admin])
+def add_user_to_blacklist():
+    form = AddUsersToBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers(form.userID.data, form.comment.data)
+        if blocked_user == BlockedUsers.query.filter_by(userID=form.userID.data).first():
+            flash('User already blocked.')
+            return redirect(url_for('.blocked_users'))
+        g.db.add(blocked_user)
+        g.db.commit()
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: You don't handle the case where there are form errors

File: mod_ci/controllers.py [hunk 17]
```diff
@@ -866,6 +867,44 @@ def show_maintenance():
...
+
+
+@mod_ci.route('/blocked_users/add', method=['POST'])
+@login_required
+@check_access_rights([Role.admin])
+def add_user_to_blacklist():
+    form = AddUsersToBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers(form.userID.data, form.comment.data)
+        if blocked_user == BlockedUsers.query.filter_by(userID=form.userID.data).first():
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: I would just check if there's already an user with that id present (see your check for deleting). Now you compare the returned object's, which can differ on comment. If that happens, you'll get a DB-error below.

File: mod_ci/controllers.py [hunk 18]
```diff
@@ -866,6 +867,44 @@ def show_maintenance():
...
+@check_access_rights([Role.admin])
+def remove_user_from_blacklist():
+    form = RemoveUsersFromBlacklist()
+    if form.validate_on_submit():
+        blocked_user = BlockedUsers.query.filter_by(userID=form.userID.data).first()
+        if blocked_user is None:
+            flash('No such user in Blacklist')
+            return redirect(url_for('.blocked_users'))
+        g.db.remove(blocked_user)
+        g.db.commit()
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: No handling of form errors ;)

File: mod_ci/controllers.py (hunk 16 above)
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: You forgot to redirect the user, so the user will see a blank page if it all goes fine.

File: mod_ci/controllers.py (hunk 18 above)
Comment (canihavesomecoffee) [reply mod_ci/controllers.py]: No redirect if all went fine.

File: run.py [hunk 19]
```diff
@@ -72,6 +73,17 @@ def install_secret_keys(application, secret_session='secret_key', secret_csrf='s
 install_secret_keys(app)
 
 
+def username(u):
+    user_id = str(u)
```
Comment (canihavesomecoffee) [inline run.py]: Is this specific cast required? a str.format below (which requires no casting) might be clearer on intent.

File: run.py [hunk 20]
```diff
@@ -72,6 +73,17 @@ def install_secret_keys(application, secret_session='secret_key', secret_csrf='s
...
+    api_url = requests.get('https://api.github.com/user/' + user_id)
```
Comment (canihavesomecoffee) [inline run.py]: What if this request fails for any reason?
